HAL: Add a get_capabilities() function to ResetReason API

fkjagodzinski opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->

<!-- 
    Please provide the following information: 

    Description of the the change (what is this fixing / adding / removing?).

    Why the change is needed (if this is fixing a reported issue please summarize what
    the issue is and add the reference. E.g. Fixes #17119).

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
    
-->
Add the `hal_reset_reason_get_capabilities()` function to the ResetReason HAL API to skip the unsupported reason values during HAL & driver tests.

Fixes #11792.

Updated tests:
* `tests-mbed_hal-reset_reason`,
* `tests-mbed_drivers-reset_reason`.

#### Impact of changes <!-- Optional -->
<!-- 
    If there are any implications for users taking this change then they must be 
    provided here. For Major PR types this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->
Extend the ResetReason HAL API with the `hal_reset_reason_get_capabilities()` function. Unsupported reason values are skipped during the greentea tests.

#### Migration actions required <!-- Optional -->
<!-- 
    This should only be applicable in Major PR types for which this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->
A default, weak implementation is provided. Every target has to override this weak implementation to provide the correct `reset_reason_capabilities_t`.

### Documentation <!-- Required -->

<!-- 
    Please provide details of any document updates required, including links to any
    related PRs against the docs repository.
    If no document updates are required please specify 'None', this at least tells us
    that this has been considered.
-->
None

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [x] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [] No Tests required for this change (E.g docs only update)
    [x] Covered by existing mbed-os tests (Greentea or Unittest)
    [] Tests / results supplied as part of this PR
    
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

<!--
    Request additional reviewers with @username or @team
-->
@jamesbeyond, @mprse

----------------------------------------------------------------------------------------------------------------

fkjagodzinski committed:
HAL: ResetReason: Add the get_capabilities function

Add the hal_reset_reason_get_capabilities() to ResetReason HAL API.
Add a default, weak implementation, that every target can override.

fkjagodzinski committed:
Tests: ResetReason: Use the get_capabilities()

Make use of the new hal_reset_reason_get_capabilities() function to skip
unsupported reset resaons during tests.

fkjagodzinski committed:
K64F: Add hal_reset_reason_get_capabilities()

0xc0170 commented on hal/mbed_compat.c:
```diff
@@ -90,4 +91,15 @@ MBED_WEAK void spi_get_capabilities(PinName ssel, bool slave, spi_capabilities_t
+// To be re-implemented in the target layer if required
+MBED_WEAK void hal_reset_reason_get_capabilities(reset_reason_capabilities_t *cap)
+{
+    cap->reasons = 1 << RESET_REASON_PIN_RESET | 1 << RESET_REASON_SOFTWARE;
```
just a suggestion, would be more readable if  `( )` around `(1 << RESET_REASON_PIN_RESET)` 

`(1 << RESET_REASON_PIN_RESET) | (1 << RESET_REASON_SOFTWARE)`

fkjagodzinski replied:
Updated as requested.

0xc0170 reviewed (commented):
Before I approve, what other targets/families should implement this (at least to check if weakly defined is good for them) ?

fkjagodzinski committed:
TOSHIBA: Add hal_reset_reason_get_capabilities()

Override the default, weak implementation of
hal_reset_reason_get_capabilities() for TMPM066, TMPM3H6, TMPM3HQ,
TMPM46B & TMPM4G9.
